UIforETW/Utility.cpp

```diff
@@ -352,75 +743,112 @@ int DeleteFiles(HWND hwnd, const std::vector<std::wstring>& paths)
+	//TODO: IFileOperation?
+	const int result = ::SHFileOperationW(&fileOp);
 
 	return result;
 }
 
 void SetClipboardText(const std::wstring& text)
 {
-	BOOL cb = OpenClipboard(GetDesktopWindow());
-	if (!cb)
+	if (!openClipboard())
 		return;
 
-	EmptyClipboard();
+	if (!emptyClipboard())
+	{
+		closeClipboard();
+		return;
+	}
 
-	size_t length = (text.size() + 1) * sizeof(wchar_t);
-	HANDLE hmem = GlobalAlloc(GMEM_MOVEABLE, length);
-	if (hmem)
+	HANDLE hmem_temp = FALSE;
+	if (!AllocGlobalMemoryForString(text, &hmem_temp))
 	{
-		void *ptr = GlobalLock(hmem);
-		if (ptr != NULL)
-		{
-			memcpy(ptr, text.c_str(), length);
-			GlobalUnlock(hmem);
+		closeClipboard();
+		return;
+	}
+	const HANDLE& hmem = hmem_temp;
 
-			SetClipboardData(CF_UNICODETEXT, hmem);
-		}
+	void* ptr_temp = nullptr;
+	if (!LockGlobalMemory(hmem, &ptr_temp))
+	{
+		FreeGlobalMemory(hmem);
+		closeClipboard();
+		return;
 	}
 
-	CloseClipboard();
+	void* const ptr = ptr_temp;
+
+	wcscpy_s(static_cast<wchar_t*>(ptr), (text.size() + 1), text.c_str());
+
+	UnlockGlobalMemory(hmem);
+
+	if (!ClipboardSetUnicode(hmem))
+	{
+		FreeGlobalMemory(hmem);
+		closeClipboard();
+		return;
+	}
+	closeClipboard();
 }
 
 std::wstring GetClipboardText()
 {
 	std::wstring result;
-	BOOL cb = OpenClipboard(GetDesktopWindow());
-	if (!cb)
+	if (!openClipboard())
 		return result;
 
-	HANDLE hClip = GetClipboardData(CF_UNICODETEXT);
-	if (hClip)
+	HANDLE hClip_temp = NULL;
+	if (!ClipboardGetUnicode(&hClip_temp))
 	{
-		wchar_t *text = static_cast<wchar_t *>(GlobalLock(hClip));
-		if (text)
-		{
-			size_t bytes = GlobalSize(hClip);
-			result.insert(result.begin(), text, text + bytes / sizeof(wchar_t));
-		}
-		::GlobalUnlock(hClip);
+		closeClipboard();
+		return result;
 	}
 
-	CloseClipboard();
+	const HANDLE hClip = hClip_temp;
 
+	void* ptr = nullptr;
+	if (!LockGlobalMemory(hClip, &ptr))
+	{
+		closeClipboard();
+		return result;
+	}
+
+	PCWSTR const text = static_cast<PCWSTR>(ptr);
+
+	size_t bytes_temp = 0u;
+	if (!SizeGlobalMemory(hClip, &bytes_temp))
+	{
+		UnlockGlobalMemory(hClip);
+		closeClipboard();
+		return result;
+	}
+
+	const size_t bytes = bytes_temp;
+
+	result.insert(result.begin(), text, text + bytes / sizeof(wchar_t));
+
+	UnlockGlobalMemory(hClip);
+	closeClipboard();
 	return result;
 }
 
 int64_t GetFileSize(const std::wstring& path)
 {
-	LARGE_INTEGER result;
-	HANDLE hFile = CreateFile(path.c_str(), GENERIC_READ,
-		FILE_SHARE_READ | FILE_SHARE_WRITE, NULL, OPEN_EXISTING,
+	LARGE_INTEGER result = {0};
+	HANDLE hFile = ::CreateFileW(path.c_str(), GENERIC_READ,
+		(FILE_SHARE_READ | FILE_SHARE_WRITE), NULL, OPEN_EXISTING,
 		FILE_ATTRIBUTE_NORMAL, NULL);
 	if (hFile == INVALID_HANDLE_VALUE)
+	{
+		debugPrintf(L"Failed to get file size!\n");
+		debugLastError();
 		return 0;
+	}
 
-	if (GetFileSizeEx(hFile, &result))
-	{
-		CloseHandle(hFile);
+	const BOOL gfsResult = ::GetFileSizeEx(hFile, &result);
+	CloseValidHandle(hFile);

```

Comment:
Probably off topic, but instead of CloseValidHandle wouldn't it be better to have a ScopedHandle class that would close the handle automatically (and do any appropriate error checking)?